godot/game/game.gd

```diff
@@ -67,12 +67,26 @@ func _process(delta):
 		if death_timer.is_stopped():
 			sound.play()
 		sec_since_last_magician_sfx = -sound.stream.get_length()
-
-	var r = randi()%(450+int(game_timer.time_left*1.5))
+	
+	environment.set_game_progress_ratio(1.-game_timer.time_left/max_game_time)
+	
+	# spawn objects
+	
+	# range used for random calculations
+	# scale with time to increase likelyhood of objects spawning near the end
+	var master_probability = 450.+game_timer.time_left*1.5
+	# increase range on higher refreshrates
+	#TODO this should ideally be the inverse multiplied onto the probabilities itself
+	master_probability = master_probability*(1./60.)/delta
+	
+	#TODO rewrite return / move into function
 	if game_timer.time_left < 9.0:
-		r = 10
 		level.kill_lasers()
-	if r > 0 and r <= 7:
+		return
+	
+	#TODO reusing rand makes for uncleaner code but is less efficient, maybe there is another approach?
+	#TODO define proper probabilities
+	if randf()*master_probability < 7:
```

Comment:
Multiplying randf() with master_probability separately for the different things (lasers, books, powerups) allows for multiple things to happen in a single frame. This shouldn't really affect game-balance, so I think this is fine. Originally, I had intended to simply have a weighted choice between the different options (laser, book, powerups or nothing at all).